### ISSUE TITLE
client::reconnect() don't work, application keeps block.

### ISSUE THREAD

**Issue by zhoushiqi88** [eclipse/paho.mqtt.cpp#300]
I developed an application with mqtt::client,but, mqtt::client::reconnect don't work, Application keeps block.connect(_connopts) is OK.

**Comment by fpagliughi** [comment]
Can you submit a sample showing what you're doing.

**Comment by danielnobs** [comment]
I can confirm the problem reported by @zhoushiqi88. The following program reproduces the bug.
Tested on Linux and Windows 10 with paho-mqtt-cpp 1.1 

```
#include <mqtt/client.h>

//
// https://github.com/eclipse/paho.mqtt.cpp/issues/300
// 

int main()
{
    const std::string TOPIC { "hello" };
    const std::string PAYLOAD1 { "hello world 1!" };
    const std::string PAYLOAD2 { "hello world 2!" };

    // Create a client
    const std::string ADDRESS = "127.0.0.1";
    const std::string CLIENT_ID = "MyId";
    const int QOS = 0;
    mqtt::client cli(ADDRESS, CLIENT_ID);

    mqtt::connect_options connOpts;
    connOpts.set_keep_alive_interval(20);
    connOpts.set_clean_session(true);
    try {
        cli.connect(connOpts);
        auto msg1 = mqtt::make_message(TOPIC, PAYLOAD1);
        msg1->set_qos(QOS);
        cli.publish(msg1);
        cli.disconnect();

        std::cerr << "trying to reconnect..."  << std::endl;
        cli.reconnect(); // reconnect blocks forever
        std::cerr << "successfully reconnected!"  << std::endl;

        auto msg2 = mqtt::make_message(TOPIC, PAYLOAD2);
        msg2->set_qos(QOS);
        cli.publish(msg2);
    }
    catch (const mqtt::exception& exc) {
        std::cerr << "Error: " << exc.what() << " [" << exc.get_reason_code() << "]" << std::endl;
        return 1;
    }

    return 0;
}

```

**Comment by fpagliughi** [comment]
OK. Thanks for the sample. I recreated the problem and am working to get it fixed.

**Comment by fpagliughi** [comment]
Thanks for this error report. I think the behavior of the underlying C library changed at some point, and I failed to catch it. But this investigation also uncovered another, even more serious, bug in that the synchronous client was not properly throwing an exception when it timed out! Both of those bugs have now been fixed and will be in the pending release. That will come out in the next few days once the C lib v1.3.8 is released.
